test/Microsoft.IdentityModel.Tokens.Tests/CryptoProviderFactoryTests.cs

```diff
@@ -817,6 +829,144 @@ public static TheoryData<CryptoProviderFactoryTheoryData> ReleaseRsaKeyWrapProvi
+
+            Assert.Null(signing.CryptoProviderCache);
+            Assert.Null(signing2.CryptoProviderCache);
+
+            cryptoProviderFactory.ReleaseSignatureProvider(signing2);
+            var disposeCalled = (bool)signing2.GetType().GetField("_disposed", BindingFlags.NonPublic | BindingFlags.Instance).GetValue(signing2);
+            Assert.False(disposeCalled);
+            cryptoProviderFactory.ReleaseSignatureProvider(signing);
+            disposeCalled = (bool)signing.GetType().GetField("_disposed", BindingFlags.NonPublic | BindingFlags.Instance).GetValue(signing);
+            Assert.True(disposeCalled);
+        }
+
+        [Fact]
+        public void ReferenceCountingTest_MultiThreaded()
+        {
+            var context = new CompareContext($"{this}.ReferenceCountingTest_MultiThreaded");
+            var cryptoProviderFactory = new CryptoProviderFactory(CryptoProviderCacheTests.CreateCacheForTesting());
+
+            Task[] tasks = new Task[1000];
+
+            for (int i = 0; i < 1000; i++)
+            {
+                tasks[i] = Task.Run(() =>
+                {
+                    var rsaSha256 = cryptoProviderFactory.CreateForSigning(Default.AsymmetricSigningKey, Default.AsymmetricSigningAlgorithm);
+                    var hmacSha256 = cryptoProviderFactory.CreateForSigning(Default.SymmetricSigningKey, SecurityAlgorithms.HmacSha256Signature);
+                    var hmacSha512 = cryptoProviderFactory.CreateForSigning(Default.SymmetricSigningKey512, ALG.HmacSha512);
+                    var hmacSha384 = cryptoProviderFactory.CreateForSigning(Default.SymmetricSigningKey384, ALG.HmacSha384);
+
+                    var rsaSha256Verifying = cryptoProviderFactory.CreateForVerifying(Default.AsymmetricSigningKey, Default.AsymmetricSigningAlgorithm);
+                    var hmacSha256Verifying = cryptoProviderFactory.CreateForVerifying(Default.SymmetricSigningKey, SecurityAlgorithms.HmacSha256Signature);
+                    var hmacSha512Verifying = cryptoProviderFactory.CreateForVerifying(Default.SymmetricSigningKey512, ALG.HmacSha512);
+                    var hmacSha384Verifying = cryptoProviderFactory.CreateForVerifying(Default.SymmetricSigningKey384, ALG.HmacSha384);
+
+                    cryptoProviderFactory.ReleaseSignatureProvider(rsaSha256);
+                    cryptoProviderFactory.ReleaseSignatureProvider(hmacSha256);
+                    cryptoProviderFactory.ReleaseSignatureProvider(hmacSha512);
+                    cryptoProviderFactory.ReleaseSignatureProvider(hmacSha384);
+
+                    cryptoProviderFactory.ReleaseSignatureProvider(rsaSha256Verifying);
+                    cryptoProviderFactory.ReleaseSignatureProvider(hmacSha256Verifying);
+                    cryptoProviderFactory.ReleaseSignatureProvider(hmacSha512Verifying);
+                    cryptoProviderFactory.ReleaseSignatureProvider(hmacSha384Verifying);
+                });
+            }
+
+            Task.WaitAll(tasks);
```

Comment:
> Task.WaitAll(tasks); [](start = 11, length = 21)

should it be asserted that right after the wait the ref count should be 0 ?

Reply:
Makes for an easier to understand test perhaps.